File: frontend/src/types/character.ts
export interface Demeanor {
  speech_style?: string;
  habits?: string;
  typical_reaction?: string;
  expressiveness?: string;
}

export interface Psyche {
  desire?: string;
  fear?: string;
  conflict?: string;
  self_perception?: string;
}
export interface Contour {
  appearance?: string;
  age_era?: string;
  identity?: string;
  first_impression?: string;
}

export interface Anchor {
  essence: string;
  name: string;
  tags: string[];
  theme?: string;
  core_belief?: string;
  summary?: string;
}


export interface Trace {
  background?: string;
  key_events: string[];
  turning_point?: string;
}

export interface Bond {
  attitude_to_others?: string;
  intimate_pattern?: string;
  hostile_pattern?: string;
  group_role?: string;
}

export interface CharacterData {
  contour: Contour;
  demeanor: Demeanor;
  psyche: Psyche;
  anchor: Anchor;
  trace: Trace;
  bond: Bond;
}


export interface ForkRequest {
  new_name?: string;
  author_id?: string;
}

export interface ForkChainItem {
  id: string;
  name: string;
  author_id: string;
  created_at: string;
  is_current: boolean;
}

export interface ForkChain {
  original_id: string;
  original_name: string;
  total_forks: number;
  chain: ForkChainItem[];
}


export interface Tag {
  id: string;
  name: string;
  color: string;
  usage_count: number;
  created_at: string;
}

export interface Character {
  id: string;
  name: string;
  author_id: string;
  character_data: CharacterData;
  is_public: boolean;
  tags: Tag[];
  image_path?: string;
  fork_from: string | null;
  created_at: string;
  updated_at: string;
}

export interface CharacterCreate {
  character_data: CharacterData;
  is_public: boolean;
  image_path?: string;
}

export interface CharacterUpdate {
  character_data?: CharacterData;
  is_public?: boolean;
  image_path?: string;
}
export interface CharacterSearchResult {
  items: Character[];
  total: number;
  skip: number;
  limit: number;
}

// ============================================
// 关系系统
// ============================================

export interface RelationResponse {
  id: string;
  source_id: string;
  target_id: string;
  target_name: string;
  target_essence?: string;
  relation_name: string;
  relation_type: string;
  description?: string;
  is_mutual: boolean;
  created_at: string;
}

export interface RelationCreate {
  target_id: string;
  relation_name: string;
  relation_type: string;
  description?: string;
  is_mutual: boolean;
}

export interface RelationGraphNode {
  id: string;
  name: string;
  essence?: string;
  is_center: boolean;
}

export interface RelationGraphLink {
  source: string;
  target: string;
  relation_name: string;
  is_mutual: boolean;
}

export interface RelationGraph {
  nodes: RelationGraphNode[];
  links: RelationGraphLink[];
}

export interface PresetCategory {
  [category: string]: PresetRelation[];
}

export interface PresetRelation {
  name: string;
  category: string;
  mutual: boolean;
}
export interface ExportOptions {
  format: "system" | "plain" | "markdown" | "json";
  include_stories: boolean;
  include_relations: boolean;
}

export interface ExportResult {
  character_data: CharacterData;
  name: string;
  tags: string[];
  stories?: { title: string; content: string; word_count: number }[];
  relations?: any[];
}

File: backend/src/app/schemas/character.py
from pydantic import BaseModel, Field, model_validator
from typing import Optional, List, Dict, Any
from uuid import UUID
from datetime import datetime
from .tag import TagResponse


# 六层结构保持不变
class Contour(BaseModel):
    appearance: Optional[str] = Field(None)
    age_era: Optional[str] = Field(None)
    identity: Optional[str] = Field(None)
    first_impression: Optional[str] = Field(None)


class Demeanor(BaseModel):
    speech_style: Optional[str] = Field(None)
    habits: Optional[str] = Field(None)
    typical_reaction: Optional[str] = Field(None)
    expressiveness: Optional[str] = Field(None)


class Psyche(BaseModel):
    desire: Optional[str] = Field(None)
    fear: Optional[str] = Field(None)
    conflict: Optional[str] = Field(None)
    self_perception: Optional[str] = Field(None)


class Anchor(BaseModel):
    essence: str = Field("", description="一句话本质概括")
    name: str = Field("", description="姓名/称谓")
    tags: List[str] = Field(default_factory=list, description="标签")
    summary: Optional[str] = Field(None, description="卡片副标题，比essence更短")
    theme: Optional[str] = Field(None)
    core_belief: Optional[str] = Field(None)


class Trace(BaseModel):
    background: Optional[str] = Field(None)
    key_events: List[str] = Field(default_factory=list)
    turning_point: Optional[str] = Field(None)


class Bond(BaseModel):
    attitude_to_others: Optional[str] = Field(None)
    intimate_pattern: Optional[str] = Field(None)
    hostile_pattern: Optional[str] = Field(None)
    group_role: Optional[str] = Field(None)


class CharacterData(BaseModel):
    contour: Contour = Field(default_factory=Contour)
    demeanor: Demeanor = Field(default_factory=Demeanor)
    psyche: Psyche = Field(default_factory=Psyche)
    anchor: Anchor = Field(default_factory=Anchor)
    trace: Trace = Field(default_factory=Trace)
    bond: Bond = Field(default_factory=Bond)


class CharacterCreate(BaseModel):
    character_data: CharacterData
    is_public: bool = False
    image_path: Optional[str] = None


class CharacterUpdate(BaseModel):
    character_data: Optional[CharacterData] = None
    is_public: Optional[bool] = None
    image_path: Optional[str] = None


class CharacterResponse(BaseModel):
    id: UUID
    name: str
    author_id: str
    character_data: Dict[str, Any]
    is_public: bool
    tags: List[TagResponse] = []
    fork_from: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    image_path: Optional[str] = None
    view_count: int = 0

    class Config:
        from_attributes = True


class CharacterForkRequest(BaseModel):
    new_name: Optional[str] = Field(None)
    author_id: str = Field("anonymous")


class CharacterSearchResponse(BaseModel):
    items: List[CharacterResponse]
    total: int
    skip: int
    limit: int


class CharacterExportRequest(BaseModel):
    format: str = Field("markdown", pattern="^(system|plain|markdown|json)$")
    include_stories: bool = False
    include_relations: bool = False
